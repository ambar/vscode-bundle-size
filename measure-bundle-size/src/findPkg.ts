import escalade from 'escalade'
import {promises as fs} from 'fs'
import {createRequire} from 'module'
import path from 'path'

export type Pkg = {
  name: string
  version: string
  description?: string
  homepage?: string
  repository?: string
  dependencies?: Record<string, string>
  peerDependencies?: Record<string, string>
}

export const findPkg = (file: string) => {
  return escalade(file, (_, names) => {
    if (names.includes('package.json')) {
      return 'package.json'
    }
  })
}

const readJson = async (file: string) => {
  const content = await fs.readFile(file, 'utf-8')
  return JSON.parse(content) as Record<string, unknown>
}

const pkgName = 'package.json'
export const findPkgByName = (
  file: string,
  moduleName: string,
  callback?: (file: string, pkg: Pkg) => void
) => {
  return escalade(file, async (dir, names) => {
    if (names.includes(pkgName)) {
      const file = path.join(dir, pkgName)
      const json = (await readJson(file)) as Pkg
      callback?.(file, json)
      if (json.name === moduleName) {
        return pkgName
      }
    }
  })
}

export const findPkgs = async (modulePath: string, baseDir: string) => {
  const contextRequire = createRequire(path.resolve(baseDir, '<import>.js'))
  // ensure resolvable
  const moduleFile = contextRequire.resolve(modulePath)
  const isScoped = modulePath[0] === '@'
  const [p1, p2] = isScoped
    ? modulePath.split('/')
    : [modulePath.split('/')[0], '']
  const moduleName = p1 + (isScoped ? '/' : '') + p2
  let rootPkg: Pkg | void
  let rootPkgFile: string | void
  try {
    rootPkgFile = `${moduleName}/package.json`
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    rootPkg = require(rootPkgFile)
  } catch (_) {
    // This is probably: Package subpath './package.json' is not defined by "exports"
  }

  let modulePkg: Pkg | void
  let modulePkgFile: string | void
  if (moduleFile !== moduleName) {
    const pairs: [string, Pkg][] = []
    const pkgFile = await findPkgByName(moduleFile, moduleName, (x, y) => {
      pairs.push([x, y])
    })
    if (pkgFile && pairs.length > 0) {
      ;[rootPkgFile, rootPkg] = pairs[pairs.length - 1]
      if (pairs.length > 1) {
        ;[modulePkgFile, modulePkg] = pairs[0]
      }
    }
  }
  return {rootPkgFile, rootPkg, modulePkgFile, modulePkg}
}
