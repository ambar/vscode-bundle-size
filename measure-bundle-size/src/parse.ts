import * as parser from '@babel/parser'
import traverse, {NodePath} from '@babel/traverse'
import * as t from '@babel/types'

type Namespace = {
  name: string
  usingProps: {
    start: number
    end: number
    name: string
  }[]
}

export type ImportInfo = {
  start: number
  end: number
  from: string
  names?: Record<string, string>
  namespace?: Namespace
}

export type ParseResult = {
  imports: ImportInfo[]
}

export const parse = (input: string): ParseResult => {
  const ast = parser.parse(input, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  })
  if (ast.errors.length) {
    throw new Error(ast.errors[0].code)
  }
  const format = (x: t.ImportDeclaration) => {
    const {start, end, specifiers, source} = x
    const info: ImportInfo = {
      start: start as number,
      end: end as number,
      from: source.value,
    }
    for (const s of specifiers) {
      if (s.type === 'ImportNamespaceSpecifier') {
        info.namespace = {name: s.local.name, usingProps: []}
      } else if (s.type === 'ImportDefaultSpecifier') {
        info.names ||= {}
        info.names.default = s.local.name
      } else if (s.type === 'ImportSpecifier') {
        info.names ||= {}
        info.names[
          s.imported.type === 'StringLiteral'
            ? s.imported.value
            : s.imported.name
        ] = s.local.name
      }
    }
    return info
  }

  // top level only
  const imports = ast.program.body
    .filter((x) => x.type === 'ImportDeclaration')
    .map((x) => format(x as t.ImportDeclaration))

  const namespaces = imports
    .map((x) => x.namespace)
    .filter(Boolean) as Namespace[]
  if (namespaces.length) {
    const names = new Set(namespaces.map((x) => x.name))
    const createVisitor =
      <T extends t.Identifier | t.JSXIdentifier>(
        type: 'MemberExpression' | 'JSXMemberExpression',
        propType: 'Identifier' | 'JSXIdentifier'
      ) =>
      (path: NodePath<T>) => {
        const {name} = path.node
        if (!names.has(name)) {
          return
        }
        const {parent} = path
        if (parent.type === type) {
          const binding = path.scope.getBinding(name)
          if (
            binding &&
            binding.path.node.type === 'ImportNamespaceSpecifier'
          ) {
            const namespace = namespaces.find(
              (e) => e.name === name
            ) as Namespace
            // ignore Expression | PrivateName
            if (parent.property.type === propType) {
              namespace.usingProps.push({
                start: parent.start as number,
                end: parent.end as number,
                name: parent.property.name,
              })
            }
          }
        }
      }
    traverse(ast, {
      Identifier: createVisitor<t.Identifier>('MemberExpression', 'Identifier'),
      JSXIdentifier: createVisitor<t.JSXIdentifier>(
        'JSXMemberExpression',
        'JSXIdentifier'
      ),
    })
  }

  return {imports}
}

export const exportImported = (info: ImportInfo) => {
  const {names, namespace} = info
  return [
    names && `export {${Object.values({...names}).join(', ')}}`,
    namespace?.usingProps.length &&
      `export default [${namespace.usingProps
        .map((x) => `${namespace.name}.${x.name}`)
        .join(', ')}]`,
  ]
    .filter(Boolean)
    .join('\n')
}
