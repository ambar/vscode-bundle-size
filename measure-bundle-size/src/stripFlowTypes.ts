import {transform} from 'sucrase'

/**
 * Strip flow types from the given source code.
 *
 * Why sucrase is used instead of babel, bundle size of this package:
 * - `sucrase`: 848kb -> 1.1mb
 * - `@babel/core` (with `@babel/plugin-transform-flow-strip-types`): 848kb -> 5.7mb
 */
export function stripFlowTypes(source: string, filename?: string) {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
  const result = transform(source, {
    transforms: ['flow', 'jsx', 'imports'],
    jsxRuntime: 'automatic',
    production: true,
    filePath: filename,
  })
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
  return result.code
}
