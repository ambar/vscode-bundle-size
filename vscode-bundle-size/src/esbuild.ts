import {exec} from 'child_process'
import {promisify} from 'util'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFunction<T = void> = (...args: any[]) => T

function safeGet<T>(fn: AnyFunction<T>, defaultValue: T) {
  try {
    return fn()
  } catch (_) {
    return defaultValue
  }
}

const hasEsbuild = () => safeGet(() => !!require.resolve('esbuild'), false)
const command = promisify(exec)

/**
 * Get the right native executable of esbuild since it cannot be bundled
 * consider downloading: https://esbuild.github.io/getting-started/#download-a-build
 */
export const install = async (onInstall: AnyFunction) => {
  if (hasEsbuild()) {
    return
  }
  onInstall()
  let err: unknown
  for (const cmd of ['pnpm i --prod', 'yarn --prod', 'npm i --prod']) {
    await command(cmd, {cwd: __dirname}).catch((e) => {
      err = e
    })
    if (hasEsbuild()) {
      return
    }
  }
  return Promise.reject(err)
}
