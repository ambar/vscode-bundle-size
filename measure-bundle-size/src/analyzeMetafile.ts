import {sep} from 'path'
import bytes from 'bytes'
import * as esbuild from 'esbuild'

const groupBy = <T, K = string>(list: T[] = [], fn: (x: T) => K) => {
  const group = new Map<K, T[]>()
  for (const x of list) {
    const key = fn(x)
    const value = group.get(key)
    if (value) {
      value.push(x)
    } else {
      group.set(key, [x])
    }
  }
  return group
}

const splitModuleName = (path = '') => {
  const [p1, p2, ...rest] = path.split(sep)
  return p1.startsWith('@')
    ? [p1 + '/' + p2, rest.join(sep)]
    : [p1, [p2, ...rest].join(sep)]
}

const npmDir = 'node_modules'
const maxFilesInCell = 10

/**
 * Like `esbuild.analyzeMetafile`, but:
 * - render to markdown table
 * - more compact, accumulate files in same module
 * - strip common ancestor directory
 */
export default function analyzeMetafile(data: esbuild.Metafile) {
  const {outputs} = data
  const markdown: string[] = []

  for (const [outputName, outputStats] of Object.entries(outputs)) {
    const {entryPoint, inputs, bytes: totalBytes} = outputStats
    const list = Object.entries(inputs).map(([x, {bytesInOutput}]) => {
      if (x === entryPoint) {
        return {names: ['', x, ''], bytesInOutput}
      }
      const index = x.indexOf(npmDir)
      const parent = x.slice(0, index)
      const [moduleName, part2] = splitModuleName(
        x.slice(index + npmDir.length + 1)
      )
      return {names: [parent, moduleName, part2], bytesInOutput}
    })
    const group = groupBy(list, (x) => x.names[0])
    const nameGrouped = [...group].map(([_, entries]) => {
      return groupBy(entries, (x) => x.names[1])
      // TODO: show parent dir if needed, alert duplicates
      // return groupBy(entries, (x) => x.names[0] + x.names[1])
    })

    const head = ['Input', 'Files', 'Size', 'Percent']
    const lines = [
      `| ${head.join(' | ')} |`,
      `| ${head.map(() => '---').join(' | ')} |`,
    ]
    const leadingRow = {
      name: outputName,
      files: '',
      size: totalBytes,
      percent: '100%',
    }
    const rows = [leadingRow]
    for (const map of nameGrouped) {
      for (const [name, group] of map) {
        const files = group
          .map((x) => ({name: x.names[2], size: x.bytesInOutput}))
          .sort((a, b) => b.size - a.size)
        const bytesAcc = group.reduce((acc, x) => acc + x.bytesInOutput, 0)
        if (bytesAcc === 0) {
          continue
        }
        const percent = ((bytesAcc / totalBytes) * 100).toFixed(1) + '%'
        const getFiles = () => {
          if (files.length === 1) {
            return files[0].name
          }
          const format = (x: typeof files[0]) => `${x.name} (${bytes(x.size)})`
          if (files.length > maxFilesInCell) {
            const rest = files.slice(maxFilesInCell)
            const restSize = rest.reduce((acc, x) => acc + x.size, 0)
            return (
              files.slice(0, maxFilesInCell).map(format).join(', ') +
              `, and ${rest.length} other ${
                rest.length > 1 ? 'files' : 'file'
              } (${bytes(restSize)})`
            )
          }
          return files.map(format).join(', ')
        }
        rows.push({
          name: 'â†ª ' + name,
          files: getFiles(),
          size: bytesAcc,
          percent,
        })
      }
    }
    const maxLen = Math.max(...rows.map((r) => r.files.length))
    if (maxLen < 10) {
      // spread table
      leadingRow.files = '&nbsp; '.repeat(10)
    }
    const code = (text: string) => `\`${text}\``
    for (const row of rows.sort((a, b) => b.size - a.size)) {
      const cells = [code(row.name), row.files, bytes(row.size), row.percent]
      lines.push(`| ${cells.join(' | ')} |`)
    }
    markdown.push(lines.join('\n'))
  }

  return markdown.join('\n\n')
}
