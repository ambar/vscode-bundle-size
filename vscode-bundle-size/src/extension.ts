import path from 'path'
import * as vscode from 'vscode'
import {format} from 'pretty-format'
import type * as mbs from 'measure-bundle-size'
import {install} from './esbuild'

const output = vscode.window.createOutputChannel('Bundle Size')

const jsLanguageIds = [
  'javascript',
  'javascriptreact',
  'typescript',
  'typescriptreact',
]

const getColor = (size: number) => {
  const config = vscode.workspace.getConfiguration('bundleSize')
  const sizeKB = size / 1024
  if (sizeKB >= config.dangerSize) {
    return config.dangerColor as string
  } else if (sizeKB >= config.cautionSize) {
    return config.cautionColor as string
  }
  return config.infoColor as string
}

const channelLog: AnyVoidFunction = (...args) => {
  // eslint-disable-next-line no-console
  console.log(...args)
  output.appendLine(
    args
      .map((x) => (typeof x === 'object' ? format(x, {min: true}) : String(x)))
      .join(' ')
  )
}
const log: AnyVoidFunction = (...args) => {
  channelLog(new Date().toJSON(), 'extension', ...args)
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyVoidFunction = (...args: any[]) => void

type DecorationInfo = {
  measure: mbs.MeasureResult
  decoration: vscode.DecorationOptions
}
const documentDecorationInfoMap = new WeakMap<
  vscode.TextDocument,
  DecorationInfo[]
>()

let installPromise: ReturnType<typeof install>
let lastWorkspaceFolder: vscode.WorkspaceFolder | void
export function activate(_context: vscode.ExtensionContext) {
  log('Bundle Size is now active!')
  installPromise = install(() => log('downloading esbuild on first use'))
  installPromise.catch(log)

  // on initial
  void processDocument(vscode.window.activeTextEditor?.document)
  // on file open/close
  vscode.window.onDidChangeActiveTextEditor((e) => {
    if (e && !e.document.isUntitled) {
      lastWorkspaceFolder = vscode.workspace.getWorkspaceFolder(e.document.uri)
    }
    void processDocument(e?.document)
  })

  // on file changes
  vscode.workspace.onDidChangeTextDocument(
    debounce((e) => {
      void processDocument(e.document)
    }, 300)
  )

  vscode.languages.registerHoverProvider(jsLanguageIds, {
    provideHover(document, position) {
      const decorationInfo = documentDecorationInfoMap.get(document)
      if (!decorationInfo) {
        return
      }
      const info = decorationInfo.find((x) =>
        x.decoration.range.contains(position)
      )
      if (info) {
        const {result, error}: mbs.MeasureResult = info.measure
        // line break or para break: https://github.com/microsoft/vscode/issues/86291#issuecomment-561841915
        const lb = '  \n'
        const pb = '\n\n'
        const block = (text: string, type = 'text') =>
          '```' + type + lb + text + lb + '```'
        if (error) {
          return {contents: [`Bundle Error:${pb} ${format(error)}`]}
        } else if (result) {
          const {pkgFile, pkg, stats} = result
          const pkgUri = pkgFile && vscode.Uri.file(path.resolve(pkgFile))
          const info = pkg
            ? [
                // headline
                [
                  pkg.name && `${pkg.name}&commat;${pkg.version}`,
                  pkgUri && `[package.json](<${String(pkgUri)}>)`,
                  pkg.homepage && `${pkg.homepage}`,
                  pkg.description &&
                    pkg.description.length < 30 &&
                    pkg.description,
                ]
                  .filter(Boolean)
                  .join(' | ') + lb,
              ]
                .filter(Boolean)
                .join('')
            : ''
          return {
            contents: [
              info,
              stats,
              pkg && block(JSON.stringify(pkg, null, '  '), 'json'),
            ].filter(Boolean) as string[],
          }
        }
      }
    },
  })
}

export function deactivate() {
  //
}

const decorationType = vscode.window.createTextEditorDecorationType({
  // prevent conflicts:
  // https://github.com/microsoft/vscode/issues/33852
  // https://github.com/eamodio/vscode-gitlens/blob/main/src/annotations/lineAnnotationController.ts#L25
  after: {margin: '0 0 0 3em'},
  rangeBehavior: vscode.DecorationRangeBehavior.ClosedOpen,
})

async function processDocument(document?: vscode.TextDocument) {
  if (!document || !jsLanguageIds.includes(document.languageId)) {
    return
  }
  const editor = vscode.window.activeTextEditor
  if (!editor) {
    return
  }
  await installPromise
  if (editor !== vscode.window.activeTextEditor) {
    return
  }

  // clear console & decoration
  output.clear()
  editor.setDecorations(decorationType, [])
  documentDecorationInfoMap.set(document, [])

  const {fileName} = document
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const {measureIterable} = require('measure-bundle-size') as {
    measureIterable: typeof mbs.measureIterable
  }
  // by default, untitled file will save to `lastWorkspaceFolder`
  const workspaceFolder =
    lastWorkspaceFolder || vscode.workspace.workspaceFolders?.[0]
  const config = vscode.workspace.getConfiguration('bundleSize')

  const decorationInfo: DecorationInfo[] = []
  const flowPattern = config.get<string>('flowPattern')
  for await (const measure of measureIterable(document.getText(), fileName, {
    cache: config.get('cache') ?? true,
    debug: true,
    stats: 'table',
    log: channelLog,
    workspaceFolder: workspaceFolder?.uri.fsPath,
    flowPattern: flowPattern ? new RegExp(flowPattern) : undefined,
  })) {
    if (editor !== vscode.window.activeTextEditor) {
      break
    }
    const {importInfo, result, error} = measure
    if (result) {
      const pos = editor.document.positionAt(importInfo.end)
      const contentText = `${result.human.size} (${result.human.zippedSize} zipped)`
      const decoration = {
        range: new vscode.Range(pos, pos),
        renderOptions: {after: {contentText, color: getColor(result.size)}},
      }
      decorationInfo.push({decoration, measure})
      editor.setDecorations(
        decorationType,
        decorationInfo.map((x) => x.decoration)
      )
      documentDecorationInfoMap.set(document, decorationInfo)
    } else {
      // show error in editor?
      if (!/Skip/.test((error as Error)?.message)) {
        log('exception', measure.importInfo.from, error)
      }
    }
  }
}

function debounce<T extends AnyVoidFunction>(fn: T, wait: number) {
  let timer: number | undefined
  return function () {
    if (timer) {
      clearTimeout(timer)
    }
    // eslint-disable-next-line prefer-rest-params
    timer = setTimeout(fn.apply.bind(fn, void 0, arguments), wait)
  } as T
}
