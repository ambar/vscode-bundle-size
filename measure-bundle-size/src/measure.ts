import {promises as fs} from 'fs'
import path from 'path'
import bytes from 'bytes'
import Debug from 'debug'
import gzipSize from 'gzip-size'
import * as esbuild from 'esbuild'
import {builtinModules} from 'module'
import {parse, ImportInfo, exportImported} from './parse'
import analyzeMetafile from './analyzeMetafile'
import {findPkg, findPkgs, Pkg} from './findPkg'

const log = Debug('measure-bundle-size')

export const hasFile = async (file: string) =>
  Boolean(await fs.stat(file).catch(() => null))

export const pick = <T>(obj: T, keys: (keyof T)[]) =>
  keys.reduce((acc, k) => ((acc[k] = obj[k]), acc), {} as T)

const reBuiltin = RegExp(`^node:|^(${builtinModules.join('|')})(/|$)`)
/** ignore nested core modules  */
const builtinExternalPlugin: esbuild.Plugin = {
  name: 'external',
  setup(build) {
    build.onResolve({filter: reBuiltin, namespace: 'file'}, (args) => ({
      path: args.path,
      external: true,
    }))
  },
}
const reNonRelative = /^[a-z@]/
/** ignore nested peer deps  */
const peerExternalPlugin: esbuild.Plugin = {
  name: 'external',
  setup(build) {
    type CacheValue = Pkg | null
    const pkgCache = new Map<string, Promise<CacheValue>>()
    const externalsMap = new Map<string, Set<string>>()
    build.onResolve(
      {filter: reNonRelative, namespace: 'file'},
      async (args) => {
        if (reBuiltin.test(args.path)) {
          return null
        }
        // should passed by esbuild: https://github.com/evanw/esbuild/issues/978
        let pkg: CacheValue
        const cacheKey = args.resolveDir
        if (pkgCache.has(cacheKey)) {
          pkg = (await pkgCache.get(cacheKey)) as CacheValue
        } else {
          const result = findPkg(cacheKey).then((pkgFile) =>
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            pkgFile ? (require(pkgFile) as Pkg) : null
          )
          pkgCache.set(cacheKey, result)
          pkg = await result
        }
        if (pkg?.peerDependencies && args.path in pkg.peerDependencies) {
          let externals = externalsMap.get(pkg.name)
          if (!externals) {
            externals = new Set()
            externalsMap.set(pkg.name, externals)
          }
          externals.add(args.path)
          return {
            path: args.path,
            external: true,
          }
        }
      }
    )
    build.onEnd(() => {
      if (externalsMap.size) {
        log(`peer externals`, externalsMap)
      }
    })
  },
}

type StatsOpt = boolean | 'tree' | 'table'

type BundleResult = {
  size: number
  zippedSize: number
  human: {size: string; zippedSize: string}
  pkg?: Pkg
  pkgFile?: string
  modulePkg?: Pkg
  modulePkgFile?: string
  stats?: string
}

const bundleCache = new Map<string, BundleResult>()
const genCacheKey = (pkg: Pkg, names: string) =>
  `${pkg.name}:${pkg.version}:${names}`

const timeMark = <T extends string>() => {
  const startTimes = new Map<T, number>()
  const values = new Map<T, number>()
  return {
    start(label: T) {
      startTimes.set(label, Date.now())
    },
    end(label: T) {
      const startTime = startTimes.get(label)
      if (!startTime) {
        throw new Error(`No label: ${label}`)
      }
      values.set(label, Date.now() - startTime)
    },
    print() {
      return [...values].map(([k, v]) => `${k} in ${v}ms`).join(', ')
    },
  }
}

const pickPkgFields = (pkg: Pkg) => {
  return pick<Pkg>(pkg, [
    'name',
    'version',
    'description',
    'homepage',
    'dependencies',
    'peerDependencies',
  ])
}

const bundle = async (
  statement: string,
  importInfo: ImportInfo,
  {
    baseDir,
    projectPkgFile,
    stats: statsOpt = false,
    cache: cacheOpt = false,
  }: {
    baseDir: string
    projectPkgFile?: string | void
    stats?: StatsOpt
    cache?: boolean
  }
): Promise<BundleResult> => {
  const bundleMark = timeMark<'bundle' | 'zip' | 'analyze'>()
  const modulePath = importInfo.from
  const [part1] = modulePath.split('/')
  if (!/^[a-z@]/.test(part1)) {
    // file path, alias
    throw new Error('Skip non-npm packages')
  }
  if (reBuiltin.test(part1)) {
    throw new Error('Skip builtin modules')
  }
  const entryInput = `${statement}\n${exportImported(importInfo)}`
  type T = Awaited<ReturnType<typeof findPkgs>>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const {rootPkgFile, rootPkg, modulePkgFile, modulePkg} = await findPkgs(
    modulePath,
    baseDir
  ).catch(() => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-return
    return {} as T
  })

  const cacheKey =
    cacheOpt && rootPkg !== undefined && genCacheKey(rootPkg, entryInput)
  if (cacheOpt && cacheKey) {
    const cache = bundleCache.get(cacheKey)
    if (cache) {
      log('using cache:', importInfo.from)
      return cache
    }
  }

  bundleMark.start('bundle')
  // TODO: alert indirect deps, eg.: react-redux > react-dom
  // exclude peer deps, eg: react-router-dom > react
  const external = rootPkg ? Object.keys({...rootPkg.peerDependencies}) : []
  // reduce depth of relative path in `metafile`, can't resolve directly to `node_modules` because it may not exist (or in monorepo)
  const workingDir = projectPkgFile ? path.dirname(projectPkgFile) : baseDir
  /**
   * - build in memory, no file writing
   * - respect 'browser' field in package.json
   * - production build, respect `NODE_ENV`
   *
   * TODO: add loader for `.node`
   * @see https://esbuild.github.io/api/#build-api
   */
  const buildResult = await esbuild.build({
    stdin: {
      loader: 'ts',
      contents: entryInput,
      resolveDir: workingDir,
      sourcefile: '<import>',
    },
    outfile: '<bundle>.js',
    absWorkingDir: workingDir,
    plugins: [builtinExternalPlugin, peerExternalPlugin],
    external,
    target: 'esnext',
    format: 'esm',
    platform: 'browser',
    metafile: statsOpt !== false,
    minify: true,
    bundle: true,
    write: false,
  })
  bundleMark.end('bundle')

  // analyze on hover?
  let stats
  if (buildResult.metafile) {
    bundleMark.start('analyze')
    if (statsOpt === 'tree') {
      stats = await esbuild.analyzeMetafile(buildResult.metafile)
    } else if (statsOpt === 'table') {
      stats = analyzeMetafile(buildResult.metafile)
    }
    bundleMark.end('analyze')
    // if (stats) {
    //   log('stats:\n' + stats)
    // }
  }
  // concat <bundle>.js, <bundle>.css, etc
  bundleMark.start('zip')
  const sizes = await Promise.all(
    buildResult.outputFiles.map(async (file) => {
      const itsSize = file.contents.length
      return [
        itsSize,
        itsSize === 0 ? 0 : await gzipSize(Buffer.from(file.contents)),
      ] as [number, number]
    })
  )
  bundleMark.end('zip')
  let size = 0
  let zippedSize = 0
  for (const [s1, s2] of sizes) {
    size += s1
    zippedSize += s2
  }
  const human = {size: bytes(size), zippedSize: bytes(zippedSize)}
  log(`${bundleMark.print()}, ${statement}, ${human.size}`)
  const toRelative = path.relative.bind(null, process.cwd())
  const result = {
    size,
    zippedSize,
    human,
    pkg: rootPkg ? pickPkgFields(rootPkg) : undefined,
    pkgFile: rootPkgFile ? toRelative(rootPkgFile) : undefined,
    modulePkg: modulePkg ? pickPkgFields(modulePkg) : undefined,
    modulePkgFile: modulePkgFile ? toRelative(modulePkgFile) : undefined,
    stats,
  }
  if (cacheOpt && cacheKey) {
    bundleCache.set(cacheKey, result)
  }
  return result
}

export type MeasureResult = {
  importInfo: ImportInfo
  result?: BundleResult
  error?: Error | unknown
}

type MeasureOptions = {
  debug?: boolean
  stats?: StatsOpt
  cache?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  log?: (...args: any[]) => void
  workspaceFolder?: string
}

export const measure = async (
  input: string,
  fileName?: string | null,
  opts: MeasureOptions = {}
): Promise<MeasureResult[]> => {
  if (opts.debug) {
    log.enabled = true
  }
  if (opts.log) {
    log.log = opts.log
  }

  log('process', fileName)
  const baseDir =
    fileName && (await hasFile(fileName))
      ? path.dirname(fileName)
      : opts.workspaceFolder && (await hasFile(opts.workspaceFolder))
      ? opts.workspaceFolder
      : null
  if (!baseDir) {
    throw new Error('Can not resolve `fileName` or `workspaceFolder`')
  }

  const projectPkgFile = await findPkg(baseDir)
  const parseMark = timeMark<'parse'>()
  parseMark.start('parse')
  const result = parse(input)
  parseMark.end('parse')
  log(parseMark.print())
  const bundleOpts = {
    baseDir,
    projectPkgFile,
    stats: opts.stats,
    cache: opts.cache,
  }
  return Promise.all(
    result.imports.map(async (importInfo) => {
      const statement = input.substring(importInfo.start, importInfo.end)
      return bundle(statement, importInfo, bundleOpts).then(
        (result) => ({
          importInfo,
          result,
          error: void 0,
        }),
        (error: Error) => ({
          importInfo,
          result: void 0,
          error,
        })
      )
    })
  )
}
