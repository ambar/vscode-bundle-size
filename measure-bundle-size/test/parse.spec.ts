import {parse, exportImported} from '../src/parse'

// https://mdn.io/import
const syntaxInput = `
import defaultExport from "module-1"
import * as name from "module-2"
import { export1 } from "module-3"
import { export1 as alias1 } from "module-4"
import { export2 , export3 } from "module-5"
import { export4 , export5 as alias2 } from "module-6"
import defaultExport2, { export6 } from "module-7"
import defaultExport3, * as name2 from "module-8"
// https://github.com/tc39/ecma262/pull/2154
import {'string literal' as name3} from "module-9"
import "module-10"
`.trim()

const namespaceInput = `
import * as name from "module";

// using
{
  if (true) {
    name.using1()
    typeof name.using2
  }
}

export let foo = name.using3
export let jsx = <name.using4 />

// redeclare: block
{
  let name = {}
  name.foo()
}

// redeclare: parent block
{
  let name = {}
  if (true) {
    name.foo()
  }
}

// redeclare: hoisted function declaration
{
  name.foo()
  function name() {}
}

// redeclare: deconstruct object
{
  let {name} = {}
  name.foo()
}

// redeclare: deconstruct object rest
{
  let {...name} = {}
  name.foo()
}

// redeclare: deconstruct array
{
  let [name] = []
  name.foo()
}

// redeclare: deconstruct array rest
{
  let [...name] = []
  name.foo()
}

// redeclare: params in FunctionDeclaration
{
  function f1(name) {
      name.foo()
  }
  function f2(...name) {
      name.foo()
  }
}

// redeclare: params in FunctionExpression
{
  void function (name) {
      name.foo()
  }
  void function (...name) {
      name.foo()
  }
}

// redeclare: params in ArrowFunctionExpression
{
  (name) => {
    name.foo()
  }
  (...name) => {
    name.foo()
  }
}

// redeclare: params in CatchClause
{
  try {
  } catch(name) {
    name.foo()
  }
}
`

test('syntax', () => {
  const result = parse(syntaxInput)
  expect(result).toMatchSnapshot('result')
  expect(result.imports.map(exportImported)).toMatchSnapshot('exportImported')
})

test('namespace', () => {
  const result = parse(namespaceInput)
  expect(result).toMatchInlineSnapshot(`
    Object {
      "imports": Array [
        Object {
          "end": 32,
          "from": "module",
          "namespace": Object {
            "name": "name",
            "usingProps": Array [
              Object {
                "end": 74,
                "name": "using1",
                "start": 63,
              },
              Object {
                "end": 99,
                "name": "using2",
                "start": 88,
              },
              Object {
                "end": 135,
                "name": "using3",
                "start": 124,
              },
              Object {
                "end": 165,
                "name": "using4",
                "start": 154,
              },
            ],
          },
          "start": 1,
        },
      ],
    }
  `)
  expect(result.imports.map(exportImported)).toMatchInlineSnapshot(`
    Array [
      "export default [name.using1, name.using2, name.using3, name.using4]",
    ]
  `)
})

test('hashbang', () => {
  expect(parse(`#!/usr/bin/env node`)).toMatchInlineSnapshot(`
    Object {
      "imports": Array [],
    }
  `)
})

test('jsx', () => {
  expect(parse(`export default <div />`)).toMatchInlineSnapshot(`
    Object {
      "imports": Array [],
    }
  `)
})

test('ts', () => {
  expect(parse(`type t = 0`)).toMatchInlineSnapshot(`
    Object {
      "imports": Array [],
    }
  `)
})
