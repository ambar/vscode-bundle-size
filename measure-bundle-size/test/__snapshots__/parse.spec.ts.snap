// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syntax: exportImported 1`] = `
[
  "export {defaultExport}",
  "",
  "export {export1}",
  "export {alias1}",
  "export {export2, export3}",
  "export {export4, alias2}",
  "export {defaultExport2, export6}",
  "export {defaultExport3}",
  "export {name3}",
  "",
]
`;

exports[`syntax: result 1`] = `
{
  "imports": [
    {
      "end": 36,
      "from": "module-1",
      "names": {
        "default": "defaultExport",
      },
      "start": 0,
    },
    {
      "end": 69,
      "from": "module-2",
      "namespace": {
        "name": "name",
        "usingProps": [],
      },
      "start": 37,
    },
    {
      "end": 104,
      "from": "module-3",
      "names": {
        "export1": "export1",
      },
      "start": 70,
    },
    {
      "end": 149,
      "from": "module-4",
      "names": {
        "export1": "alias1",
      },
      "start": 105,
    },
    {
      "end": 194,
      "from": "module-5",
      "names": {
        "export2": "export2",
        "export3": "export3",
      },
      "start": 150,
    },
    {
      "end": 249,
      "from": "module-6",
      "names": {
        "export4": "export4",
        "export5": "alias2",
      },
      "start": 195,
    },
    {
      "end": 300,
      "from": "module-7",
      "names": {
        "default": "defaultExport2",
        "export6": "export6",
      },
      "start": 250,
    },
    {
      "end": 350,
      "from": "module-8",
      "names": {
        "default": "defaultExport3",
      },
      "namespace": {
        "name": "name2",
        "usingProps": [],
      },
      "start": 301,
    },
    {
      "end": 446,
      "from": "module-9",
      "names": {
        "string literal": "name3",
      },
      "start": 396,
    },
    {
      "end": 465,
      "from": "module-10",
      "start": 447,
    },
  ],
}
`;
